### ISSUE TITLE
[Bug] Android resolver fails to copy srcaar files to Maven repository for Packages in custom external paths

### ISSUE THREAD

**Issue by jonsenson-uken** [googlesamples/unity-jar-resolver#701]
### [REQUIRED] Please fill in the following fields:

  * Unity editor version: `2022.3.22`
  * External Dependency Manager version: `1.2.180`
  * Source you installed EDM4U: `.unitypackage`
  * Features in External Dependency Manager in use: `Android Resolver`
  * Plugins SDK in use: `Firebase, Facebook`
  * Platform you are using the Unity editor on: `Mac`

### [REQUIRED] Please describe the issue here:
(Please list the full steps to reproduce the issue. Include device logs, Unity logs, and stack traces if available.)

Packages that are are not in either the `Packages` or `Assets` folder generate an exception when attempting to copy .srcaar files into the local Maven repo.  These are Packages referenced using local file paths as described in the [Unity docs here](https://docs.unity3d.com/Manual/upm-localpath.html).  In our case our Firebase package lives in a folder parallel to `Assets` so the manifest entry looks like:

```
"dependencies": {
  "com.company.firebase.app": "file:../CompanyPackages/node_modules/com.company.firebase.app",
  etc...
}
```

The exception happens in [FileUtils.CreateFolder](https://github.com/googlesamples/unity-jar-resolver/blob/master/source/VersionHandlerImpl/src/FileUtils.cs#L642-L650).  Since an external folder path is sent into this method and it only checks to see if the folder exists in `AssetDatabase`, it gets stuck in a recursive loop until `path` is an empty string and `new DirectoryInfo(path)` throws the following exception:

```
Job failed with exception: System.ArgumentException: The specified path is not of a legal form (empty).
  at System.IO.Path.InsecureGetFullPath (System.String path) [0x00025] in <b904252b6b4e4277834bcca7e51f318d>:0 
  at System.IO.Path.GetFullPath (System.String path) [0x00000] in <b904252b6b4e4277834bcca7e51f318d>:0 
  at System.IO.DirectoryInfo..ctor (System.String path) [0x00006] in <b904252b6b4e4277834bcca7e51f318d>:0 
  at (wrapper remoting-invoke-with-check) System.IO.DirectoryInfo..ctor(string)
  at Google.FileUtils.CreateFolder (System.String path, Google.Logger logger) [0x00010] in <35cc61ad4ecc44e4bdeaad3e377694d3>:0 
  at Google.FileUtils.CreateFolder (System.String path, Google.Logger logger) [0x0001e] in <35cc61ad4ecc44e4bdeaad3e377694d3>:0 
  at Google.FileUtils.CreateFolder (System.String path, Google.Logger logger) [0x0001e] in <35cc61ad4ecc44e4bdeaad3e377694d3>:0 
  at Google.FileUtils.CreateFolder (System.String path, Google.Logger logger) [0x0001e] in <35cc61ad4ecc44e4bdeaad3e377694d3>:0 
  at Google.FileUtils.CreateFolder (System.String path, Google.Logger logger) [0x0001e] in <35cc61ad4ecc44e4bdeaad3e377694d3>:0 
  at Google.FileUtils.CreateFolder (System.String path, Google.Logger logger) [0x0001e] in <35cc61ad4ecc44e4bdeaad3e377694d3>:0 
  at Google.FileUtils.CreateFolder (System.String path, Google.Logger logger) [0x0001e] in <35cc61ad4ecc44e4bdeaad3e377694d3>:0 
  at Google.FileUtils.CreateFolder (System.String path, Google.Logger logger) [0x0001e] in <35cc61ad4ecc44e4bdeaad3e377694d3>:0 
  at Google.FileUtils.CreateFolder (System.String path, Google.Logger logger) [0x0001e] in <35cc61ad4ecc44e4bdeaad3e377694d3>:0 
  at Google.FileUtils.CreateFolder (System.String path, Google.Logger logger) [0x0001e] in <35cc61ad4ecc44e4bdeaad3e377694d3>:0 
  at Google.FileUtils.CreateFolder (System.String path, Google.Logger logger) [0x0001e] in <35cc61ad4ecc44e4bdeaad3e377694d3>:0 
  at GooglePlayServices.PlayServicesResolver.CopyAssetAndLabel (System.String sourceLocation, System.String targetLocation, System.Boolean force) [0x0009b] in <bd68bc0fc2b34107a6e5a63d15643d80>:0 
  at GooglePlayServices.GradleTemplateResolver.CopySrcAars (System.Collections.Generic.ICollection`1[T] dependencies) [0x000be] in <bd68bc0fc2b34107a6e5a63d15643d80>:0 
  at GooglePlayServices.GradleTemplateResolver.InjectDependencies (System.Collections.Generic.ICollection`1[T] dependencies) [0x00140] in <bd68bc0fc2b34107a6e5a63d15643d80>:0 
  at GooglePlayServices.PlayServicesResolver+<>c__DisplayClass115_0.<ResolveUnsafe>b__3 () [0x0000c] in <bd68bc0fc2b34107a6e5a63d15643d80>:0 
  at Google.RunOnMainThread.ExecuteNext () [0x00047] in <35cc61ad4ecc44e4bdeaad3e377694d3>:0 
UnityEngine.Debug:LogError (object)
Google.RunOnMainThread:ExecuteNext ()
Google.RunOnMainThread/<>c:<ExecuteAllUnnested>b__28_0 ()
Google.RunOnMainThread:RunAction (System.Action)
Google.RunOnMainThread:ExecuteAllUnnested (bool)
Google.RunOnMainThread:Run (System.Action,bool)
GooglePlayServices.PlayServicesResolver:ResolveUnsafe (System.Action`1<bool>,bool,bool)
GooglePlayServices.PlayServicesResolver:ResolveUnsafeAfterPromptCheck (System.Action`1<bool>,bool,bool,bool)
GooglePlayServices.PlayServicesResolver/<>c__DisplayClass113_0:<ResolveUnsafeAfterJetifierCheck>b__0 (bool)
GooglePlayServices.PlayServicesResolver/<>c__DisplayClass144_0:<CanEnableJetifierOrPromptUser>b__1 (GooglePlayServices.PlayServicesResolver/ApiLevelJetifierResult)
GooglePlayServices.PlayServicesResolver:CheckApiLevelForJetifier (bool,string,System.Action`1<GooglePlayServices.PlayServicesResolver/ApiLevelJetifierResult>)
GooglePlayServices.PlayServicesResolver/<>c__DisplayClass144_0:<CanEnableJetifierOrPromptUser>b__0 (bool)
GooglePlayServices.PlayServicesResolver:CheckGradleVersionForJetifier (bool,string,System.Action`1<bool>)
GooglePlayServices.PlayServicesResolver:CanEnableJetifierOrPromptUser (string,System.Action`1<bool>)
GooglePlayServices.PlayServicesResolver:ResolveUnsafeAfterJetifierCheck (System.Action`1<bool>,bool,bool,bool)
GooglePlayServices.PlayServicesResolver:ResolveUnsafeAfterMainTemplateCheck (System.Action`1<bool>,bool,bool,bool)
GooglePlayServices.PlayServicesResolver/<>c__DisplayClass110_0:<ScheduleResolve>b__1 ()
GooglePlayServices.PlayServicesResolver:ExecuteNextResolveJob ()
GooglePlayServices.PlayServicesResolver:ScheduleResolve (bool,bool,System.Action`1<bool>,bool)
GooglePlayServices.PlayServicesResolver:ExecuteMenuResolve (bool)
GooglePlayServices.PlayServicesResolver:MenuForceResolve ()
```

#### Please answer the following, if applicable:
What's the issue repro rate? 100%

What happened? How can we make the problem occur?
- Link any Package that contains a `.srcaar` file, like Firebase, in manifest.json via a relative or absolute path to any folder outside of either `Packages` or `Assets`
- Set the resolver settings to patch the main gradle Template file and enable the local Maven repo
- Run the resolver
- After the exception is triggered the resolver becomes completely unresponsive and selecting `Resolve` again does nothing at all until Unity is fully relaunched.

**Comment by ferretnt** [comment]
I've just run into this and am shocked that the only reference to this issue is a bot saying "meh, try harder."

**Comment by ferretnt** [comment]
If this was the wrong place or procedure for this bug report originally I’d love to know the correct one. Kudos to the original author for taking the time to debug.

**Comment by argzdev** [comment]
Hi folks, I took a look into this, and so far I'm unable to reproduce the same behavior. Here's the steps I took:
1. Use [Firebase Quickstart Unity for Analytics](https://github.com/firebase/quickstart-unity)
2. Open the app using Unity Editor `2022.3.40f1`
3. Add the dependencies using [UPM](https://firebase.google.com/docs/unity/setup-alternative#alternative_unity_package_manager)
    - I added the `.tgz` files in parallel with the `Packages` or `Assets` folder. (.e.g `"com.google.external-dependency-manager": "file:../GooglePackages/com.google.external-dependency-manager-1.2.182.tgz"`)
    - I checked the `patch mainTemplate.gradle` is turned on by default
    - I enabled `Use Full Custom Local Maven Repo Path > When building Android app through Unity`

So far, I tried using both methods via Package Manager UI and manifest.json, both seems to be working correctly. One thing I noticed is that you don't have the `.tgz` file extension at the end of dependency declaration. Could this be causing the issue? (`"com.company.firebase.app": "file:../CompanyPackages/node_modules/com.company.firebase.app"`). Also this is by a long shot and I'm sure it wouldn't make a difference, but could you try updating to the latest Unity version `2022.3.40f1` and see if that changes anything?

**Comment by ferretnt** [comment]
Sounds like you were able to reproduce this in the other issue #697, and that it was Unity 6 specific. If so we can probably link and close this issue.

**Comment by jonsenson-uken** [comment]
> One thing I noticed is that you don't have the .tgz file extension at the end of dependency declaration. Could this be causing the issue? ("com.company.firebase.app": "file:../CompanyPackages/node_modules/com.company.firebase.app")

`com.company.firebase.app` is actually a (unfortunately named since it clashes with mac os naming) folder containing the uncompressed contents of the library.  I'm not the one who implemented that so am unsure why or how this structure was set up.  It makes sense that linking just a whole `.tgz` file would work since the dependency manager wouldn't find the loose `.srcaar` files inside and attempt to copy them so you've effectively circumvented the bug.  The external library needs to be a raw folder that contains loose `.srcaar` files to trigger the issue.

Unfortunately we've made significant edits and progress on the project since reporting this that would make it difficult to revert and put things back into the bugged state and I don't have the free time to investigate further.  Updating to Unity 2022.3.40 to test also wouldn't be doable at this point since it contains breaking changes to our build system.

> Sounds like you were able to reproduce this in the other issue https://github.com/googlesamples/unity-jar-resolver/issues/697, and that it was Unity 6 specific, which honestly I should have confirmed, but I saw the 2022.3 at the top of the issue and so didn't. If so we can probably link and close this issue

From what I see in that issue it appears to generate the same error but the root cause in that issue is a different Unity 6 specific problem that's unrelated to this one.

**Comment by argzdev** [comment]
Thank you for the well detailed explanation, @jonsenson-uken. I'm now able to reproduce the same behavior when linking the uncompressed library. 

@ferretnt, I agree with @jonsenson-uken, this could be a different issue. But thank you for the feedback! In any case, I'll raise this to our engineering team to take a closer look and see what we can do here.

**Comment by argzdev** [comment]
Hey folks, I believe the referenced PR fixes this issue, and will be released in version `1.2.183`. That said, I'll go ahead and close this thread. Thanks!
